### PR TITLE
Feature/reservation form

### PR DESCRIPTION
**Pull request by themonster2015**
This PR adds the following to `dev`:

- Create Reservation Form
- Create Routing to reservation form (which is at `/reserve`) in App component
- Install moment.js and react-datepicker packages to work with date fields on the form.
- Add styling to reservation form so that background image changes on destination selection and reselection.
- Destination field will be disabled by default if a user arrives at the form after clicking the Reserve button from a destination detail page. The destination field will display a list of available destinations to choose from if the user arrives at the form from the main home page.
Screen Shot 2021-11-21 at 19 52 27

**Mobile**
Screen Shot 2021-11-21 at 19 52 47

### CODE REVIEW

**Review by AleksandarIvezic — CHANGES_REQUESTED** [review]
Hi @themonster2015 ,

Can you please check the error I am getting? I can't render the page because of that.

Cheers!!! 🥂

File: src/components/Reservation.js [hunk 1]
```diff
@@ -1,9 +1,187 @@
...
+  //     image_url: 'https://res.cloudinary.com/dqdrsilxz/image/upload/v1636573679/sample.jpg',
+  //     price_per_day: '100',
+  //   },
+  // ];
+
+  const setBackgroundImg = () => {
+    const url = destinations.filter((place) => place.name === chosenDestination);
+    return url[0].image_url;
+  };
+
```
Comment (AleksandarIvezic) [inline src/components/Reservation.js]: Hey Yen there is some issue with this peace of code. I can't render the form.
image

Comment (themonster2015) [reply src/components/Reservation.js]: I've commented out the part for changing background. You can check it again.

File: src/components/Reservation.js [hunk 2]
```diff
@@ -1,9 +1,187 @@
...
+  const [submitted, setSubmitted] = useState(false);
+  let { destination } = props;
+  const destinationPassed = destination !== 'default';
+  destination = destination === 'default' ? 'Bora Bora' : destination;
+  const [chosenDestination, setChosenDestination] = useState(destination);
+  const dispatch = useDispatch();
+
+  const username = useSelector((state) => state.user.username);
+  const token = useSelector((state) => state.user.token);
+
```
Comment (AleksandarIvezic) [inline src/components/Reservation.js]: Hey Yen your username is undefined. I believe that is the reason why I can't see the username auto filled. 
When I modify this peace of code
image

I am getting this in console 
image

**Review by AleksandarIvezic — CHANGES_REQUESTED** [review]
Hi @themonster2015 ,
Now I can render the form, but I can't see the username auto-filled.

image


I believe you should check your useSelector() function for defining username. Please check the comment below.

Cheers!!! 🥂

File: src/components/Reservation.js (hunk 2 above)
Comment (themonster2015) [reply src/components/Reservation.js]: Hi @ShinobiWarior thanks for pointing it out. I tried to fix it. But after that, right now I have the problem of the authentication token expired on the server already when the post request gets sent.

File: src/components/App.js [hunk 3]
```diff
@@ -23,7 +25,22 @@ const App = () => {
...
+        )}
+          />
+          <Route
+            path="/reserve"
+            element={(
+              <PrivateRoute redirectTo="/login">
+                <Reservation />
+              </PrivateRoute>
+        )}
+          />
```
Comment (AleksandarIvezic) [inline src/components/App.js]: Why instead this combination of Route and PrivateRoute components you don't use just Route? Something like this: 
```
<Route path="/home" element={<Home />} />
<Route path="/reserve" element={<Reservation />} />
```
In my case this solved the error I was getting earlier.

**Review by AleksandarIvezic — CHANGES_REQUESTED** [review]
Hi @themonster2015 , 

Nice job. The layout looks very good, but I have small issues running the app locally.

When I log in myself the app is trying to redirect me to the home page and I am getting this error.
image

The same result appears when I try to follow the /reserve path. 

I have a simple solution. Please check the comments below.

**Comment by AleksandarIvezic** [comment]
Also, I forget to mention, that it will be nice to use another font. I think that  `'Montserrat', sans-serif` would be appropriate.

File: src/components/App.js (hunk 3 above)
Comment (themonster2015) [reply src/components/App.js]: thanks for pointing it out @ShinobiWarior. I will fix it as you suggest.

**Comment by AleksandarIvezic** [comment]
@themonster2015 , 
Just noticed one more thing 
image
Please follow the naming conventions when creating CSS files. 

Cheers!!! 🥂

File: src/components/Reservation.js [hunk 4]
```diff
@@ -1,9 +1,238 @@
...
+  useEffect(() => {
+    const res = isTokenExpired();
+    if (res) {
+      localStorage.removeItem('token');
+      dispatch(logOutUser());
+      navigate('/login');
+    }
+    dispatch(clearStatus());
+  }, []);
+
```
Comment (AleksandarIvezic) [inline src/components/Reservation.js:43]: Hey @themonster2015 ,
I just figured out that this hook makes my side to redirect to the login page, and considering that I am logged in it redirects me to the home page each time I try to use the  `/reserve` path. That's mean that your i`sTokenExpired` function returns true, in spite it should return false. Can you fix it, please?

File: src/styles/reservation.module.css [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    linear-gradient(rgba(150, 191, 1, 0.8), rgba(150, 191, 1, 0.8)),
+    url("https://images.unsplash.com/photo-1601871925712-3c803d50fa72?ixid=MnwxMjA3fDB8MHxzZWFyY2h8MXx8YmVhY2glMjB2YWNhdGlvbnxlbnwwfHwwfHw%3D&ixlib=rb-1.2.1&w=1000&q=80");
+  background-repeat: no-repeat;
+  background-position: center center;
+  background-size: cover;
+  margin: auto auto;
+  padding: 40px;
+  border-radius: 5px;
+  box-shadow: 0 0 10px #000;
+  width: 90%;
```
Comment (AleksandarIvezic) [inline src/styles/reservation.module.css]: [Optional]
Hi @themonster2015 , 
You can just remove the line18 and 22, so it can look better in mobile view.

**Review by AleksandarIvezic — APPROVED** [review]
## Hi @themonster2015 ,

### It works now. Great job. You can merge it. ✔️ ✔️ ✔️ 

Before merge please make sure to check the comment below.

Cheers!!! 🥂

**Pull request merged**